### DISCUSSION TITLE
Post-render qmd document

### DISCUSSION THREAD

**Discussion by JosiahParry** [quarto-dev/quarto-cli#8019] [Q&A]
### Description

Related to https://github.com/quarto-dev/quarto-r/issues/62 which does not seem to have a follow up discussion. 

The docs for post-render fall under project scripts https://quarto.org/docs/projects/scripts.html#pre-and-post-render. Is it possible to use these features on standalone documents? I assume not but want to be sure. For example the below silently ignores the `post-render` rather than provide a warning or an error.

````
---
title: "Can you post-render a document"
post-render:
  - post-render.R
---

This should run a post render script
````

```r
# this should be printed to stdout
message(Sys.getenv("QUARTO_PROJECT_OUTPUT_FILES"))

# this should create a new file called out.txt
writeLines("hello world", "out.txt")
```

**Comment by mcanouil** [comment]
Pre and post render are project options thus only at project levels.
This is stated in the first sentence of the section.
Maybe in the future, a document will have a behaviour closer to projects, but this won't happen soon.

Note that Quarto cannot really warn for "invalid" frontmatter as you can have custom things in it.
Edit: there is some YAML validation plan or at least items to be improved which are labelled as https://github.com/quarto-dev/quarto-cli/labels/yaml-validation

**Reply by cscheid** [reply]
This is just to confirm that we _will_ have single-file documents behaving closer to projects, but not for 1.5. It's a matter of when, not if.

And we're 100% on agreement on issuing a warning on options that are valid/invalid in particular contexts. It turns out to be a tricky thing to do, but keep an eye out on https://github.com/quarto-dev/quarto-cli/labels/lint as well.

**Reply by Nova-Scotia** [reply]
Any updates on this @cscheid ? I'm trying with a single document to get post-render to work and struggling. @JosiahParry have you had any luck? I'm trying to post-render an R script to copy and rename my output html doc.

**Reply by JosiahParry** [reply]
No, not for standalone. What you can do is use the R package and do something like

```r
outfp <- “myname.html”
quarto::render(“in.qmd”, output_file = outfp)
```

**Reply by Nova-Scotia** [reply]
Thanks @JosiahParry . I'm writing a script for a friend and trying to have everything be in one place (one parameterized quarto doc). Well, there will be other files that migrate with the project, but he'll only have to pay attention to the .qmd file. 

Right after I wrote that comment, I finally figured out a working solution. It's... very... hacky. I want my quarto doc to post-render an R script named "Rename_monthly_maps_file.R". 

In my .qmd file, the YAML contains:
```
project:
  post-render: Rename_monthly_maps_file.R #ensure the _quarto.yml file is in the project directory
```
and in the .qmd file, I create a text file with the preferred filename for my output html, like this:
```
save_filename <- str_c(gsub(" ", "_", params$project), 
                       format(date_start, "%Y%m%d"), "to", 
                       format(date_end, "%Y%m%d"), sep = "_")
write_lines(save_filename, "data/save_filename.txt")
```

then, I create a _quarto.yml file that contains this text:

```
project:
  type: default
  post-render: Rename_monthly_maps_file.R
```

finally, my Rename_monthly_maps_file.R script takes the preferred output filename from the text file, copies the html from the main folder, and places it in the maps folder.

```
new_filename <- readLines("data/save_filename.txt")

file.copy("General_monthly_maps.html",
          paste0("maps/", new_filename, ".html"))
```

**Reply by cscheid** [reply]
(When you have spare time and are willing to read a personal perspective on "any updates on this?", I've written this admittedly personal account: https://bsky.app/profile/did:plc:ls5433sf4bxpw53d47y7f2of/post/3ls25wg6mnk2y)

**Reply by Nova-Scotia** [reply]
@cscheid wow! I never saw it that way. Thanks for sharing your perspective. As a non-developer, I had no idea :).

**Reply by cscheid** [reply]
Thanks for reading - I know it's strange to see it from the other side, so I genuinely appreciate the kindness.

In any case, [here's an issue that tracks the feature request](https://github.com/quarto-dev/quarto-cli/issues/13032). We'll post there when we make progress.

**Reply by JosiahParry** [reply]
I about one of those a month and they make me 🥲 i can only imagine the feeling with the veracity you get it! 
we appreciate all of the work y'all pour into this! <3
